{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "emotional-horror",
   "metadata": {},
   "source": [
    "Sitas notebookas paima alignintus sgenus, sgenuose yra bruksneliu -, kuriuos reikia panaikinti\n",
    "1. atmetam tuos kurie neprasideda ATG pradzios kodonu\n",
    "2. minusiukus (outlier'ius) šaliname 2 būdais: \n",
    "  1. jei pozicijoje dominuoja minusiukai - šaliname sekas kurios maišo, dedam i blacklist\n",
    "  2. jei dominuoja ne minusiukai - užpildome koncensusu ir transliuojam nukleotida i proteina, kad patikrintume ar istacius raide pagal konsensusa neatsirado pabaigos kodonas viduryje geno\n",
    "3. rezultate gauna \"ncbi_sgene_good_unique_aligned_cleaned.csv\" kuriame yra sgenai be bruksneliu\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "another-louisville",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:17.534308Z",
     "start_time": "2021-05-15T16:59:17.240427Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "restricted-installation",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:17.619172Z",
     "start_time": "2021-05-15T16:59:17.535906Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping 2_059f461871887e99ab81c93d6be95c63\n",
      "Skipping 1_25080f7a80094f94da370709b0d34695\n",
      "Skipping 1_07011ff250ef4a4ace2ef9b43d5c2ba1\n",
      "Skipping 6_21591cfd60c885e07d7f57962f053be0\n",
      "Skipping 1_1c35868a205decc006af1af9332de520\n",
      "Skipping 1_bc8364434f1c405884ffb19232630d57\n",
      "Skipping 1_4cfad326ce4e88375a4940a9d520f1dd\n",
      "Skipping 1_5770e81e98ccef6a23b405da967dfb8d\n",
      "Skipping 1_1dbcc614d9587e62d666ced83f7d67c1\n",
      "Skipping 1_ca6f6b227c6eef0951de1e28652b7e2d\n",
      "Skipping 1_83783011e3066d4938cced16296c2604\n",
      "Skipping 5_0e5599a7c4545edf989e6bb8ded0bcc8\n",
      "Skipping 1_73ab13457a0970f920022ef0d2e15058\n",
      "Skipping 1_370e85eb954793235737752b3be9ff8d\n"
     ]
    }
   ],
   "source": [
    "blacklist = [\n",
    "    \"1_07011ff250ef4a4ace2ef9b43d5c2ba1\", # pradzios kodonas ne ATG\n",
    "    \"1_bc8364434f1c405884ffb19232630d57\", # pradzios kodonas ne ATG\n",
    "    \"1_4cfad326ce4e88375a4940a9d520f1dd\", # pradzios kodonas ne ATG\n",
    "    \n",
    "    \"1_25080f7a80094f94da370709b0d34695\", # minusiukai nuo 822 iki 830\n",
    "    \n",
    "    \"6_21591cfd60c885e07d7f57962f053be0\", # minusiukai nuo 672 iki 680\n",
    "    \"1_370e85eb954793235737752b3be9ff8d\", # minusiukai nuo 672 iki 680\n",
    "    \n",
    "    \"1_73ab13457a0970f920022ef0d2e15058\", # minusiukai nuo 797 iki 799\n",
    "    \n",
    "    \"1_5770e81e98ccef6a23b405da967dfb8d\", # minusiukai nuo 0 iki 26\n",
    "    \"1_ca6f6b227c6eef0951de1e28652b7e2d\", # minusiukai nuo 0 iki 26\n",
    "    \"1_83783011e3066d4938cced16296c2604\", # minusiukai nuo 0 iki 26\n",
    "    \"5_0e5599a7c4545edf989e6bb8ded0bcc8\", # minusiukai nuo 0 iki 26\n",
    "    \n",
    "    \"2_059f461871887e99ab81c93d6be95c63\", # minusiukai nuo 34 iki 36\n",
    "    \"1_1c35868a205decc006af1af9332de520\", # minusiukai nuo 34 iki 36\n",
    "    \"1_1dbcc614d9587e62d666ced83f7d67c1\"  # minusiukai nuo 34 iki 36\n",
    "]\n",
    "\n",
    "def read_fasta(filename):\n",
    "    sequences = []\n",
    "    path = Path(filename)\n",
    "    with open(path, \"r\") as fasta_file:\n",
    "        sequence = {}\n",
    "        accession = None\n",
    "        for line in fasta_file:\n",
    "            if '>' in line:\n",
    "                if sequence and accession:\n",
    "                    \n",
    "                    if accession in blacklist:\n",
    "                        print(f\"Skipping {accession}\")\n",
    "                    else:\n",
    "                        sequences.append({\n",
    "                            \"accession\": accession,\n",
    "                            \"sgene_nucleotide\": \"\".join(sequence)\n",
    "                        })\n",
    "                    \n",
    "                parts = line.split(\" \")\n",
    "                accession = parts[0][1:].strip()\n",
    "                sequence = []\n",
    "            else:\n",
    "                sequence.append(line.strip())\n",
    "    \n",
    "    df = pd.DataFrame.from_dict(sequences)\n",
    "#     df.to_csv(path.with_suffix(\".csv\"), index=False)\n",
    "    return df\n",
    "\n",
    "df = read_fasta(\"good_unique.aln\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "solid-electron",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:17.643120Z",
     "start_time": "2021-05-15T16:59:17.623739Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>sgene_nucleotide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_87b59e0087b7cfc624758ac0f7cc0f89</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_7d7d2f39b893ea3ca706781d1ca03d70</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_600ed6f06199b27156f02df015050fa3</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_543814ce04c1c7813ef42e7fe0abef00</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_87472661071a21fef79f2166f4e8e631</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4364</th>\n",
       "      <td>1_5321f55a779924f878d9a69ffe18b4c1</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4365</th>\n",
       "      <td>2_fe3511bb655ecf8318fa9aee6b39a4fb</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4366</th>\n",
       "      <td>2_f3c41465d64388a1246062814595c060</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4367</th>\n",
       "      <td>1_a7434325322e04336922df03d2033a33</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4368</th>\n",
       "      <td>3_f5f66f7beb5e5bf4bb286fbce886d70e</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4369 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               accession  \\\n",
       "0     1_87b59e0087b7cfc624758ac0f7cc0f89   \n",
       "1     1_7d7d2f39b893ea3ca706781d1ca03d70   \n",
       "2     1_600ed6f06199b27156f02df015050fa3   \n",
       "3     1_543814ce04c1c7813ef42e7fe0abef00   \n",
       "4     1_87472661071a21fef79f2166f4e8e631   \n",
       "...                                  ...   \n",
       "4364  1_5321f55a779924f878d9a69ffe18b4c1   \n",
       "4365  2_fe3511bb655ecf8318fa9aee6b39a4fb   \n",
       "4366  2_f3c41465d64388a1246062814595c060   \n",
       "4367  1_a7434325322e04336922df03d2033a33   \n",
       "4368  3_f5f66f7beb5e5bf4bb286fbce886d70e   \n",
       "\n",
       "                                       sgene_nucleotide  \n",
       "0     ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "1     ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "2     ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "3     ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "4     ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "...                                                 ...  \n",
       "4364  ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "4365  ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "4366  ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "4367  ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "4368  ---------------------------ATGTTTG---TTTTTCTTG...  \n",
       "\n",
       "[4369 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "israeli-saturday",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:21.637264Z",
     "start_time": "2021-05-15T16:59:17.645574Z"
    }
   },
   "outputs": [],
   "source": [
    "# Verciam nucleotide seko i series, kad butu kiekvienai raidei po stulpeli\n",
    "def convert(s):\n",
    "    return pd.Series(list(s))\n",
    "\n",
    "_df = df[\"sgene_nucleotide\"].apply(convert)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "danish-customs",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:21.753711Z",
     "start_time": "2021-05-15T16:59:21.640915Z"
    }
   },
   "outputs": [],
   "source": [
    "df = pd.concat([df, _df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "boring-bachelor",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:21.794360Z",
     "start_time": "2021-05-15T16:59:21.755182Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>sgene_nucleotide</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>...</th>\n",
       "      <th>3863</th>\n",
       "      <th>3864</th>\n",
       "      <th>3865</th>\n",
       "      <th>3866</th>\n",
       "      <th>3867</th>\n",
       "      <th>3868</th>\n",
       "      <th>3869</th>\n",
       "      <th>3870</th>\n",
       "      <th>3871</th>\n",
       "      <th>3872</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_87b59e0087b7cfc624758ac0f7cc0f89</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_7d7d2f39b893ea3ca706781d1ca03d70</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_600ed6f06199b27156f02df015050fa3</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_543814ce04c1c7813ef42e7fe0abef00</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_87472661071a21fef79f2166f4e8e631</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4364</th>\n",
       "      <td>1_5321f55a779924f878d9a69ffe18b4c1</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4365</th>\n",
       "      <td>2_fe3511bb655ecf8318fa9aee6b39a4fb</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4366</th>\n",
       "      <td>2_f3c41465d64388a1246062814595c060</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4367</th>\n",
       "      <td>1_a7434325322e04336922df03d2033a33</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4368</th>\n",
       "      <td>3_f5f66f7beb5e5bf4bb286fbce886d70e</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4369 rows × 3875 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               accession  \\\n",
       "0     1_87b59e0087b7cfc624758ac0f7cc0f89   \n",
       "1     1_7d7d2f39b893ea3ca706781d1ca03d70   \n",
       "2     1_600ed6f06199b27156f02df015050fa3   \n",
       "3     1_543814ce04c1c7813ef42e7fe0abef00   \n",
       "4     1_87472661071a21fef79f2166f4e8e631   \n",
       "...                                  ...   \n",
       "4364  1_5321f55a779924f878d9a69ffe18b4c1   \n",
       "4365  2_fe3511bb655ecf8318fa9aee6b39a4fb   \n",
       "4366  2_f3c41465d64388a1246062814595c060   \n",
       "4367  1_a7434325322e04336922df03d2033a33   \n",
       "4368  3_f5f66f7beb5e5bf4bb286fbce886d70e   \n",
       "\n",
       "                                       sgene_nucleotide  0  1  2  3  4  5  6  \\\n",
       "0     ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "1     ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "2     ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "3     ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "4     ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "...                                                 ... .. .. .. .. .. .. ..   \n",
       "4364  ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "4365  ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "4366  ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "4367  ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "4368  ---------------------------ATGTTTG---TTTTTCTTG...  -  -  -  -  -  -  -   \n",
       "\n",
       "      7  ... 3863 3864 3865 3866 3867 3868 3869 3870 3871 3872  \n",
       "0     -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "1     -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "2     -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "3     -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "4     -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "...  ..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  \n",
       "4364  -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "4365  -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "4366  -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "4367  -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "4368  -  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "\n",
       "[4369 rows x 3875 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "handled-studio",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:24.717001Z",
     "start_time": "2021-05-15T16:59:21.797366Z"
    }
   },
   "outputs": [],
   "source": [
    "df.to_csv(\"frequencies_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "corporate-volume",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.207962Z",
     "start_time": "2021-05-15T16:59:24.718471Z"
    }
   },
   "outputs": [],
   "source": [
    "frequencies = _df.apply(pd.value_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "convertible-blackjack",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.213943Z",
     "start_time": "2021-05-15T16:59:28.209206Z"
    }
   },
   "outputs": [],
   "source": [
    "frequencies = frequencies.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "appropriate-aircraft",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.261572Z",
     "start_time": "2021-05-15T16:59:28.215902Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3868</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3869</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4368.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3870</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4369.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3871</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4369.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3872</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4368.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3873 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           -       A       C    G       T\n",
       "0     4369.0     0.0     0.0  0.0     0.0\n",
       "1     4369.0     0.0     0.0  0.0     0.0\n",
       "2     4369.0     0.0     0.0  0.0     0.0\n",
       "3     4369.0     0.0     0.0  0.0     0.0\n",
       "4     4369.0     0.0     0.0  0.0     0.0\n",
       "...      ...     ...     ...  ...     ...\n",
       "3868     0.0     0.0  4369.0  0.0     0.0\n",
       "3869     0.0  4368.0     1.0  0.0     0.0\n",
       "3870     0.0     0.0     0.0  0.0  4369.0\n",
       "3871     0.0  4369.0     0.0  0.0     0.0\n",
       "3872     0.0  4368.0     0.0  1.0     0.0\n",
       "\n",
       "[3873 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frequencies = frequencies.fillna(0)\n",
    "frequencies[frequencies['-'] > 0].to_csv(\"frequences.csv\")\n",
    "frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "reflected-crest",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.274979Z",
     "start_time": "2021-05-15T16:59:28.266089Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        True\n",
       "1        True\n",
       "2        True\n",
       "3        True\n",
       "4        True\n",
       "        ...  \n",
       "3868    False\n",
       "3869    False\n",
       "3870    False\n",
       "3871    False\n",
       "3872    False\n",
       "Length: 3873, dtype: bool"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stulpeliu maske kuriuose yra tik minusiukai\n",
    "mask_only_dash = (frequencies['-'] > 0) & (frequencies['A'] == 0)  & (frequencies['T'] == 0)  & (frequencies['G'] == 0) & (frequencies['C'] == 0)\n",
    "mask_only_dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "gorgeous-suspension",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.288935Z",
     "start_time": "2021-05-15T16:59:28.277019Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,\n",
       "        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,\n",
       "        26,  34,  35,  36, 672, 673, 674, 675, 676, 677, 678, 679, 680,\n",
       "       797, 798, 799, 822, 823, 824, 825, 826, 827, 828, 829, 830])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stulpeliu vardai kuriuose yra tik minusiukai\n",
    "columns_only_dash = mask_only_dash[mask_only_dash].index.values\n",
    "columns_only_dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "empty-security",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.743295Z",
     "start_time": "2021-05-15T16:59:28.291307Z"
    }
   },
   "outputs": [],
   "source": [
    "# Metam stulpelius kuriuose yra tik minusiukai\n",
    "df.drop(columns_only_dash, axis=1, inplace=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "revised-surrey",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.761942Z",
     "start_time": "2021-05-15T16:59:28.744814Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>sgene_nucleotide</th>\n",
       "      <th>27</th>\n",
       "      <th>28</th>\n",
       "      <th>29</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>37</th>\n",
       "      <th>...</th>\n",
       "      <th>3863</th>\n",
       "      <th>3864</th>\n",
       "      <th>3865</th>\n",
       "      <th>3866</th>\n",
       "      <th>3867</th>\n",
       "      <th>3868</th>\n",
       "      <th>3869</th>\n",
       "      <th>3870</th>\n",
       "      <th>3871</th>\n",
       "      <th>3872</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_87b59e0087b7cfc624758ac0f7cc0f89</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_7d7d2f39b893ea3ca706781d1ca03d70</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_600ed6f06199b27156f02df015050fa3</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_543814ce04c1c7813ef42e7fe0abef00</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_87472661071a21fef79f2166f4e8e631</td>\n",
       "      <td>---------------------------ATGTTTG---TTTTTCTTG...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3824 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            accession  \\\n",
       "0  1_87b59e0087b7cfc624758ac0f7cc0f89   \n",
       "1  1_7d7d2f39b893ea3ca706781d1ca03d70   \n",
       "2  1_600ed6f06199b27156f02df015050fa3   \n",
       "3  1_543814ce04c1c7813ef42e7fe0abef00   \n",
       "4  1_87472661071a21fef79f2166f4e8e631   \n",
       "\n",
       "                                    sgene_nucleotide 27 28 29 30 31 32 33 37  \\\n",
       "0  ---------------------------ATGTTTG---TTTTTCTTG...  A  T  G  T  T  T  G  T   \n",
       "1  ---------------------------ATGTTTG---TTTTTCTTG...  A  T  G  T  T  T  G  T   \n",
       "2  ---------------------------ATGTTTG---TTTTTCTTG...  A  T  G  T  T  T  G  T   \n",
       "3  ---------------------------ATGTTTG---TTTTTCTTG...  A  T  G  T  T  T  G  T   \n",
       "4  ---------------------------ATGTTTG---TTTTTCTTG...  A  T  G  T  T  T  G  T   \n",
       "\n",
       "   ... 3863 3864 3865 3866 3867 3868 3869 3870 3871 3872  \n",
       "0  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "1  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "2  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "3  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "4  ...    T    T    A    C    A    C    A    T    A    A  \n",
       "\n",
       "[5 rows x 3824 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "soviet-albert",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T15:36:08.065965Z",
     "start_time": "2021-05-15T15:36:08.050252Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fifth-wireless",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:28.772811Z",
     "start_time": "2021-05-15T16:59:28.763853Z"
    }
   },
   "outputs": [],
   "source": [
    "frames = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                 \n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'\n",
    "}\n",
    "\n",
    "def translate(seq):\n",
    "    \n",
    "    seq = seq.upper()\n",
    "    \n",
    "    protein = []\n",
    "    n = len(seq)\n",
    "    if n % 3 != 0:\n",
    "        print(\"nukleotidas nedalus is 3\")\n",
    "        return\n",
    "    \n",
    "    for i in range(0, n, 3):\n",
    "        codon = seq[i:i + 3]\n",
    "        frame = frames[codon]\n",
    "        protein.append(frame)\n",
    "        \n",
    "    return \"\".join(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "devoted-generic",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:35.570234Z",
     "start_time": "2021-05-15T16:59:28.774322Z"
    }
   },
   "outputs": [],
   "source": [
    "# pakeiciam minusiukus i dazniausiai pasitaikancias reiksmes toje pozicijoje \n",
    "# ir patikrinam ar tas genas neturi stop kodonu viduryje\n",
    "\n",
    "sgene_nucleotide_new = {}\n",
    "rows_with_modified_sgene = []\n",
    "for row_index, row in df.iterrows():\n",
    "    nucleotide = []\n",
    "    contains_dash = False \n",
    "    \n",
    "    for column_name, cell in row.iteritems():\n",
    "        if column_name in ['accession', 'sgene_nucleotide']:\n",
    "            continue\n",
    "            \n",
    "        if cell == '-':\n",
    "            contains_dash = True\n",
    "            cell = frequencies.iloc[column_name].idxmax()\n",
    "#             print(f\"Replacing - with {new_cell} at {column_name}\")\n",
    "        nucleotide.append(cell)\n",
    "    \n",
    "    nucleotide_str = \"\".join(nucleotide)\n",
    "    if contains_dash:\n",
    "        # check codons\n",
    "#         print(\"translating nucleotide to protein\")\n",
    "        rows_with_modified_sgene.append(row_index)\n",
    "        protein = translate(nucleotide_str)\n",
    "        \n",
    "        if protein is None:\n",
    "            # nucleotidas nedalus is 3\n",
    "            continue\n",
    "            \n",
    "        stop = protein.index(\"_\") # stop kodonas\n",
    "        if stop is None:\n",
    "            print(\"nera stop kodono\")\n",
    "            continue\n",
    "            \n",
    "        if len(protein) - stop > 5:\n",
    "            print(\"stop kodonas ne gale\")\n",
    "            continue\n",
    "            \n",
    "    sgene_nucleotide_new[row_index] = nucleotide_str\n",
    "# sgene_nucleotide_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "italian-portugal",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:35.579321Z",
     "start_time": "2021-05-15T16:59:35.571423Z"
    }
   },
   "outputs": [],
   "source": [
    "df['sgene_nucleotide_new'] = pd.Series(sgene_nucleotide_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "trained-direction",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:35.841357Z",
     "start_time": "2021-05-15T16:59:35.581547Z"
    }
   },
   "outputs": [],
   "source": [
    "df_new = df[['accession', 'sgene_nucleotide_new']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "subject-synthetic",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:35.854860Z",
     "start_time": "2021-05-15T16:59:35.843024Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>sgene_nucleotide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_87b59e0087b7cfc624758ac0f7cc0f89</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_7d7d2f39b893ea3ca706781d1ca03d70</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_600ed6f06199b27156f02df015050fa3</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_543814ce04c1c7813ef42e7fe0abef00</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_87472661071a21fef79f2166f4e8e631</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4364</th>\n",
       "      <td>1_5321f55a779924f878d9a69ffe18b4c1</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4365</th>\n",
       "      <td>2_fe3511bb655ecf8318fa9aee6b39a4fb</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4366</th>\n",
       "      <td>2_f3c41465d64388a1246062814595c060</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4367</th>\n",
       "      <td>1_a7434325322e04336922df03d2033a33</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4368</th>\n",
       "      <td>3_f5f66f7beb5e5bf4bb286fbce886d70e</td>\n",
       "      <td>ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4369 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               accession  \\\n",
       "0     1_87b59e0087b7cfc624758ac0f7cc0f89   \n",
       "1     1_7d7d2f39b893ea3ca706781d1ca03d70   \n",
       "2     1_600ed6f06199b27156f02df015050fa3   \n",
       "3     1_543814ce04c1c7813ef42e7fe0abef00   \n",
       "4     1_87472661071a21fef79f2166f4e8e631   \n",
       "...                                  ...   \n",
       "4364  1_5321f55a779924f878d9a69ffe18b4c1   \n",
       "4365  2_fe3511bb655ecf8318fa9aee6b39a4fb   \n",
       "4366  2_f3c41465d64388a1246062814595c060   \n",
       "4367  1_a7434325322e04336922df03d2033a33   \n",
       "4368  3_f5f66f7beb5e5bf4bb286fbce886d70e   \n",
       "\n",
       "                                       sgene_nucleotide  \n",
       "0     ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "1     ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "2     ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "3     ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "4     ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "...                                                 ...  \n",
       "4364  ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "4365  ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "4366  ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "4367  ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "4368  ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTG...  \n",
       "\n",
       "[4369 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new = df_new.rename(columns={\"sgene_nucleotide_new\": \"sgene_nucleotide\"})\n",
    "df_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "approved-failing",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-15T16:59:36.334984Z",
     "start_time": "2021-05-15T16:59:35.856455Z"
    }
   },
   "outputs": [],
   "source": [
    "df_new.to_csv(\"ncbi_sgene_good_unique_aligned_cleaned.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "brown-norway",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"ncbi_sgene_good_unique_aligned_cleaned.fasta\", \"w\") as fasta_file:\n",
    "    for index, row in df_new.iterrows():\n",
    "        fasta_file.write(f\">{row['accession']}\\n\")\n",
    "        fasta_file.write(f\"{row['sgene_nucleotide']}\\n\\n\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ignored-duncan",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
